src/modules/ChannelList/dux/reducers.js

```diff
@@ -114,79 +124,98 @@ export default function reducer(state, action) {
-          allChannels: state.allChannels.filter(({ url }) => url !== channel?.url),
-        };
       }
-      const filteredChannels = state.allChannels.filter((c) => !(c.url === channel?.url && isMe));
-      const nextChannel = (isMe && (channel?.url === state.currentChannel?.url))
-        ? filteredChannels[0]
-        : state.currentChannel;
+      // Replace the currentChannel if I left the currentChannel
+      if (isMe && channel.url === currentChannel?.url) {
+        if (!disableAutoSelect && allChannels.length > 0) {
+          const [firstChannel, secondChannel = null] = allChannels;
+          nextChannel = firstChannel.url === channel.url ? secondChannel : firstChannel;
+        }
+      } else {
+        nextChannel = currentChannel;
+      }
       return {
         ...state,
-        currentChannel: state.disableAutoSelect ? null : nextChannel,
-        allChannels: filteredChannels,
+        currentChannel: nextChannel,
+        allChannels: nextChannels,
       };
     }
     case actions.ON_USER_JOINED:
     case actions.ON_CHANNEL_CHANGED:
     case actions.ON_READ_RECEIPT_UPDATED:
     case actions.ON_DELIVERY_RECEIPT_UPDATED: {
-      const { allChannels = [] } = state;
       const channel = action.payload;
+      const {
+        allChannels = [],
+        currentUserId,
+        currentChannel,
+        channelListQuery,
+        disableAutoSelect,
+      } = state;
       const { unreadMessageCount } = channel;
-      if (!channel?.lastMessage) return state;
-      if (state.channelListQuery) {
-        if (filterChannelListParams(state.channelListQuery, channel, state.currentUserId)) {
+
+      // Do not display the channel when it's created (and not sent a message yet)
+      if (action.type === actions.ON_USER_JOINED && !channel?.lastMessage) return state;
+
+      if (channelListQuery) {
+        if (filterChannelListParams(channelListQuery, channel, currentUserId)) {
+          // Good to [add to/keep in] the ChannelList
           return {
             ...state,
             allChannels: getChannelsWithUpsertedChannel(allChannels, channel),
           };
         }
-        const nextChannel = (channel?.url === state?.currentChannel?.url)
-          ? state.allChannels[state.allChannels[0].url === channel?.url ? 1 : 0]
-          // if coming channel is first of channel list, current channel will be the next one
-          : state.currentChannel;
+        // Filter the channel from the ChannelList
+        // Replace the currentChannel if it's filtered channel
+        const filteredChannel = channel;
+        let nextChannel = null;
+        if (currentChannel?.url === filteredChannel.url) {
+          if (!disableAutoSelect && allChannels.length > 0) {
+            const [firstChannel, secondChannel = null] = allChannels;
+            nextChannel = firstChannel.url === filteredChannel.url ? secondChannel : firstChannel;
+          }
+        } else {
+          nextChannel = currentChannel;
+        }
```

Comment:
I see this logic is used in three different places. Can we wrap it into a funciton?

Reply:
good suggestion, let me check the feasibility